File: src/main/java/xyz/phanta/cbtweaker/structure/block/StructureBlockMatch.java
package xyz.phanta.cbtweaker.structure.block;

import xyz.phanta.cbtweaker.hatch.HatchTileEntity;

import javax.annotation.Nullable;

public abstract class StructureBlockMatch {

    public static StructureBlockMatch maybeComponent(@Nullable String compId) {
        return compId != null ? new Component(compId) : Normal.INSTANCE;
    }

    private StructureBlockMatch() {
        // NO-OP
    }

    public static class Normal extends StructureBlockMatch {

        public static final Normal INSTANCE = new Normal();

        private Normal() {
            // NO-OP
        }

    }

    public static class Hatch extends StructureBlockMatch {

        private final String groupId;
        private final HatchTileEntity hatch;

        public Hatch(String groupId, HatchTileEntity hatch) {
            this.groupId = groupId;
            this.hatch = hatch;
        }

        public String getGroupId() {
            return groupId;
        }

        public HatchTileEntity getHatch() {
            return hatch;
        }

    }

    public static class Component extends StructureBlockMatch {

        private final String compId;

        public Component(String compId) {
            this.compId = compId;
        }

        public String getComponentId() {
            return compId;
        }

    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/structure/block/StructureBlockMatch.kt
package st.evening.mc.cbtweaker.structure.block

import st.evening.mc.cbtweaker.hatch.HatchTileEntity
import st.evening.mc.prelude.api.data.ser.SerializationException
import st.evening.mc.prelude.api.data.tjson.JsonPath
import st.evening.mc.prelude.api.data.tjson.TJson
import st.evening.mc.prelude.api.data.tjson.expectInt
import st.evening.mc.prelude.api.data.tjson.expectStringValue

sealed interface StructureBlockMatch {
    object Normal : StructureBlockMatch

    data class Hatch(val groupId: String, val hatch: HatchTileEntity) : StructureBlockMatch

    data class Component(val componentId: String, val count: Int) : StructureBlockMatch {
        companion object {
            context(_: JsonPath)
            fun load(dto: TJson): Component = when (dto) {
                is TJson.String -> Component(dto.value, 1)
                is TJson.Object -> Component(
                    dto.expectStringValue("id"),
                    dto.expectInt("count") ?: 1
                )
                else -> throw SerializationException.withPath("Expected a component ID or an object!")
            }
        }
    }
}
